### ISSUE TITLE
meta: Mocha's place in the merged Node.js/JS Foundation

### ISSUE THREAD

**Issue by boneskull** [mochajs/mocha#3778]
(The foundation is subsequently referred to as _$FOUNDATION_.)

The following is from @jorydotcom, posted here with permission (with slight formatting tweaks by me).

Mocha is currently in the JSF's "incubator" stage, which is somewhat analogous to the proposed "growth" stage, as explained below.

* * * 

We're getting to the point in our merge conversation wherein I need to develop a picture for which projects will be coming in from the JSF at what stage. The approach I'm advocating for is to have each project lead decide for themselves what stage is the best fit. 

In a nutshell, the stages are:
- Emeritus - for projects that are functionally 'done' and expect no more contributions (though they will continue to receive basic services)
- At Large - for projects that are happy where they are in terms of activity and resource support from the foundation (I think most JSF projects will probably elect this)
- Growth - for projects that have identified growth goals, and want active foundation support to reach them
- Impact - for projects that have achieved enterprise adoption, community contributions and open governance, and have developed a self-sustaining LTS cycle 

Below the line is an excerpt that explains the stages a bit more (for your convenience) or you can view the staging document directly on github. If you don't think this is clear, you have feedback on the wording, etc - please do let me know here or on the repo. 

Regardless of the stage you choose, Mocha will be able to participate in the CPC, and potentially elect someone to serve as CPC director. Also, you can move from one stage to another - this is not a permanent selection. For example, maybe you are At Large this year, but next year you have some new features you want to implement so you move to Growth stage. Or your Emeritus project is getting a resurgence of interest from a new group of maintainers, so it makes sense for it to move to At Large. It's the circle of Open Source Life.

I wanted to get this information to you so that you can share it with your community & discuss. When you're ready, please let me know which stage you'd like to opt-in to. Of course if you need help figuring that out I am here to support you, and we will definitely cover this at the all-project meeting on Monday, 2/18. I would like to get your selection in by end of next week if that's possible. 

Best,
Jory

* * *

### At Large Projects (formerly 'Sandbox')
**Definition** 
At Large projects are projects which the CPC believes are, or have the potential to be, important to the ecosystem of Top-Level Projects or the JS ecosystem as a whole. They may be early-stage projects just getting started, or they may be long-established projects with minimal resource needs. The At Large stage provides a beneficial, neutral home for these projects in order to foster collaborative development and provide a path to deeper alignment with other _$FOUNDATION_ projects via the graduation process.

**Examples**
1. New projects that are designed to extend one or more _$FOUNDATION_ projects with functionality or interoperability libraries. 
1. Independent projects that fit the _$FOUNDATION_ mission and provide potential for a novel approach to existing functional areas (or are an attempt to meet an unfulfilled need).
1. Projects commissioned or sanctioned by the _$FOUNDATION_.
1. Any project that realistically intends to join the _$FOUNDATION_ Incubation or Top Level Stages in the future and wishes to lay the foundations for that transition.

**Expectations**
End users should evaluate At Large projects with care, as this stage does not set requirements for community size, governance, or production readiness. At Large projects will receive minimal marketing support from the _$FOUNDATION_. Projects will be reviewed on an annual basis; they may also request a status review by submitting a report to the CPC.

**Acceptance Criteria**
To be considered for the At Large Stage, the project must meet the following requirements:
* 2 CPC sponsors to champion the project & provide mentorship as needed
* A presentation to at the meeting of the CPC
* Adherence to the _$FOUNDATION_ IP Policy
* Upon acceptance, At Large projects must list their status prominently on website/readme

### Growth Stage (formerly 'Incubating')
**Definition** 
The Growth Stage is for projects that are interested in reaching the Impact Stage, and have identified a growth plan for doing so. Growth Stage projects will receive mentorship from the CPC and are expected to actively develop their community of contributors, governance, project documentation, and other variables identified in the growth plan that factor in to broad success and adoption.

In order to support their active development, projects in the Growth stage have a higher level of access to marketing and other resources, which will be agreed upon and reviewed on a yearly basis. A project's progress toward its growth plan goals will be reviewed on a yearly basis, and the CPC may ask the project to move to the At Large stage if progress on the plan drops off or stalls. 

**Examples**
1. Projects that are on their way or very likely to become Impact Projects.
1. Projects that have developed new growth targets or other community metrics for success.
1. Projects that are looking to create a lifecycle plan (maintainership succession, contributor programs, version planning, etc.)
1. Projects that need more active support from the Foundation in the form of marketing or CPC mentorship in order to reach their goals. 

**Expectations**
Projects in the Growth Stage are generally expected to move out of the Growth stage within two years. Depending on their growth plans, projects may cycle through At Large, Growth, or Impact stage as needed.  

**Acceptance Criteria**
To be considered for Growth Stage, the project must meet the At Large requirements as well as the following:

 * Development of a growth plan, to be done in conjunction with their project mentor(s) at the CPC.
 * Document that it is being used successfully in production by at least two independent end users which, in the CPC’s judgement, are of adequate quality and scope.
 * Demonstrate a substantial ongoing flow of commits and merged contributions.
 * Demonstrate that the current level of community participation is sufficient to meet the goals outlined in the growth plan.
 * Since these metrics can vary significantly depending on the type, scope and size of a project, the CPC has final judgement over the level of activity that is adequate to meet these criteria.
 * Receive a two-thirds supermajority vote of the CPC to move to Growth Stage. 

### Impact Stage (formerly 'Top-Level')
**Definition**
The Impact Stage is for projects that have reached their growth goals and are now on a sustaining cycle of development, maintainence, and long-term support. Impact Stage projects are used commonly in enterprise production environments and have large, well-established project communities.    

**Examples**
1. Projects that have publicly documented release cycles and plans for LTS.
1. Projects that have themselves become platforms for other projects.
1. Projects that are able to attract a healthy number of committers on the basis of its production usefulness (not simply 'developer popularity').
1. Projects that have several, high-profile or well known end-user implementations.

**Expectations**
Impact Stage projects are expected to participate actively in CPC proceedings, and as such have a binding vote on CPC matters requiring a formal vote, such as the election of a CPC Director. They receive ongoing financial and marketing support from the Foundation, and are expected to cross promote the foundation along with their activities. 

**Acceptance Criteria**
To graduate from At Large or Growth status, or for a new project to join as an Impact project, a project must meet the Growth stage criteria plus:

 * Have a defined governing body of at least 5 or more members (owners and core maintainers), of which no more than 1/3 is affiliated with the same employer. In the case there are 5 governing members, 2 may be from the same employer. 
 * Have a documented and publicly accessible description of the project's governance, decision-making, and release processes.
 * Have a healthy number of committers from at least two organizations. A committer is defined as someone with the commit bit; i.e., someone who can accept contributions to some or all of the project.
 * Adopt the _$FOUNDATION_ Code of Conduct.
 * Explicitly define a project governance and committer process. This is preferably laid out in a GOVERNANCE.md file and references a CONTRIBUTING.md and OWNERS.md file showing the current and emeritus committers.
 * Have a public list of project adopters for at least the primary repo (e.g., ADOPTERS.md or logos on the project website).
 * Other metrics as defined by the applying Project during the application process in cooperation with the CPC.
 * Receive a supermajority vote from the CPC to move to Impact stage. Projects can move directly from At Large to Impact, if they can demonstrate sufficient maturity and have met all requirements. 

### Emeritus Stage
**Definition**
Emeritus projects are projects which the maintainers feel have reached or are nearing end-of-life. Emeritus projects have contributed to the ecosystem, but are not necessarily recommended for modern development as there may be more actively maintained choices. The _$FOUNDATION_ appreciates the contributions of these projects and their communities, and the role they have played in moving the ecosystem forward. 

**Examples**
1. Projects that are "complete" by the maintainers' standards.
1. Projects that do not plan to release major versions in the future.

**Expectations**
Projects in this stage are not in active development. Their maintainers may infrequently monitor their repositories, and may only push updates to address security issues, if at all. Emeritus projects should clearly state their status and what any user or contributor should expect in terms of response or support. If there is an alternative project the maintainers recommend, it should be listed as well. The foundation will continue to hold the IP and any trademarks and domains, but the project does not draw on foundation resources.

**Comment by boneskull** [comment]
Mocha ticks most of the boxes for an "impact" project.  Except this:

> * Projects that have publicly documented release cycles and plans for LTS.

And we also don't have a list of "adopters".

While supporting multiple major release lines of Mocha would be helpful, we would need guidance and/or more maintenance resources to accomplish this.

I'm thinking, though, a release cycle currently doesn't fit Mocha's needs. Mocha's  development ebbs and flows.

Then, if Mocha's project goals don't necessarily align with an "impact" project, they don't necessarily align with a "growth" project either, since the point of "growth" projects is to become "impact" projects.

That leaves "at large", which doesn't sit too well with me, given that I think we could use more foundation resources, and have historically received next-to-nothing from the JSF.

**Comment by juergba** [comment]
My thoughts in a brain-storming manner:

1. We don't know anything:
We are getting offered a chair for some event, somewhere, at sometime and are hitten by conceptual definitions. What is the purpose of that organisation, of our membership? What do they expect from us? What do we get from them?
2. Benefits for Mocha:
Money? Developper power? will Mocha be part of Node's core? Do we get famous?
3. Costs for Mocha:
Which rules do we have to obey? How much time do we have to invest? in talking, writing documentation, concepts, in different languages ...
4. Our maintenance
Persuing a goal means: things have to be done, some of them are fun, some are less fun, some are total unfun. They all have to be done, with coordination and priorities.
5. Your role ...
is the bottleneck of the our actual maintenance.

**Comment by boneskull** [comment]
@juergba 	I'll try my best to answer your questions.

> 1. We don't know anything:
>     We are getting offered a chair for some event, somewhere, at sometime and are hitten by conceptual definitions. What is the purpose of that organisation, of our membership? What do they expect from us? What do we get from them?

It sounds like this is a hypothetical question?  If we need to speak about the foundation, it behooves us to learn more about it.  Mocha is of course a member of the [JS Foundation](http://js.foundation), and much of that still applies here.  See [the foundation bootstrap repo](https://github.com/nodejs/bootstrap) for further information.

Mocha *will* be a member of the merged foundation, for mostly legal reasons.  But we can choose where we stand in it.

> 2. Benefits for Mocha:
>    Money? Developper power? will Mocha be part of Node's core? Do we get famous?

Part of membership in the foundation means there will theoretically be some budget for stuff, yes.  Mocha will not be part of Node.js core, as Node.js is its own project.  I don't know how to get famous. :wink:

> 3. Costs for Mocha:
>    Which rules do we have to obey? How much time do we have to invest? in talking, writing documentation, concepts, in different languages ...

We do as little or as much as we want.  To reach, say, "impact" level, more is expected of our project, however, and we must sustain that.

There are some basic rules, such as having a code of conduct, which we're already following, since those are also part of the responsibilities of a JS Foundation member project.

> 4. Our maintenance
>    Persuing a goal means: things have to be done, some of them are fun, some are less fun, some are total unfun. They all have to be done, with coordination and priorities.

mm, sort of, but maybe not in the way you're thinking.  Node.js itself has no roadmap, and this works fine.  Plenty of coordination happens--lots of meetings and groups working towards common goals.  But nobody is anybody's boss.  Nobody gets assigned tasks.  If you don't want to do something, don't do it.  Somebody else might, if they want to.

> 5. Your role ...
>    is the bottleneck of the our actual maintenance.

I think what you're saying is tangential to this issue; let's talk about that in contributors' chat.

**Comment by boneskull** [comment]
@mochajs/core 	If you have no opinion, either I'll decide, or the foundation will decide for us.  

To distill that wall of text, what we're basically trying to do is establish if we want/expect more resources from the foundation than we're currently getting.  If that's true, then we also must accept increased requirements on our time and energy.

**Comment by craigtaub** [comment]
My thoughts from reading the wall of text

**At large**
We will continue to receive no funding. Boo.

**Growth**
- no point aiming for this unless we decide to aim for impact

**Impact**
- shame no flexibility on the "documented release cycle and plans for LTS" but I understand why
- The question is are major operation changes for the sake of more resources worth it? At this point doesn't feel like it is (to me, interested to hear others thoughts). 

Perhaps we should just go for "at large" for now as can always move forward a stage in the future if we take a more "impact" like direction.

**Comment by jorydotcom** [comment]
Hi @boneskull @juergba + Mocha Team 😄 

Just wandering in to introduce myself + answer questions. I've been contracted by the JSF specifically to advocate for projects + maintainers through the merge. I'm here to help you with anything project or foundation related and make sure you get answers from the JS or Linux Foundations as needed. You can ask me anything, at anytime, for any reason. Like your Foundation Alexa, but hopefully more useful!

> 
> > 1. We don't know anything:
> >    We are getting offered a chair for some event, somewhere, at sometime and are hitten by conceptual definitions. What is the purpose of that organisation, of our membership? What do they expect from us? What do we get from them?
> 
> It sounds like this is a hypothetical question? If we need to speak about the foundation, it behooves us to learn more about it. Mocha is of course a member of the [JS Foundation](http://js.foundation), and much of that still applies here. See [the foundation bootstrap repo](https://github.com/nodejs/bootstrap) for further information.
> 
> Mocha _will_ be a member of the merged foundation, for mostly legal reasons. But we can choose where we stand in it.
> 

It may be better to think of this a clarification of the "rules of engagement" for projects in the foundation. In the JSF's current model, those rules are unclear (what is a graduated project vs a mentored project?) and frankly the process isn't really used. This new process strives to make it clear what projects can do and should expect given different stages _that they opt-in to_.  As @boneskull  says, Mocha is a JSF project, meaning at the most basic level JSF is the legal entity holding/protecting/liable for the IP. To date, it would be fair to say you haven't really seen much benefit from that, other than the invisible benefit of risk mitigation. What additional benefits you want to draw from the new foundation are up to you, and may depend on the stage you elect. I can clarify that more if you would like.

> > 1. Benefits for Mocha:
> >    Money? Developper power? will Mocha be part of Node's core? Do we get famous?
> 
> Part of membership in the foundation means there will theoretically be some budget for stuff, yes. Mocha will not be part of Node.js core, as Node.js is its own project. I don't know how to get famous. 😉
> 

Potentially money, if the project wants to seek funds for specific efforts. But also access to things like a travel fund, cross-project events and collaboration opportunities, other foundation programs. 

> > 1. Costs for Mocha:
> >    Which rules do we have to obey? How much time do we have to invest? in talking, writing documentation, concepts, in different languages ...
> 
> We do as little or as much as we want. To reach, say, "impact" level, more is expected of our project, however, and we must sustain that.
> 
> There are some basic rules, such as having a code of conduct, which we're already following, since those are also part of the responsibilities of a JS Foundation member project.

Exactly as @boneskull says - it's up to you. The 'cost' is your time - Impact Level projects have more of an expectation on them in terms of the time they will put in to foundation-wide interests/activities. 

> 
> > 1. Our maintenance
> >    Persuing a goal means: things have to be done, some of them are fun, some are less fun, some are total unfun. They all have to be done, with coordination and priorities.
> 
> mm, sort of, but maybe not in the way you're thinking. Node.js itself has no roadmap, and this works fine. Plenty of coordination happens--lots of meetings and groups working towards common goals. But nobody is anybody's boss. Nobody gets assigned tasks. If you don't want to do something, don't do it. Somebody else might, if they want to.
> 

You can pursue goals on whatever timeframe you want to accomplish them. The Foundation may be able to help with some of the unfun bits!

**Comment by juergba** [comment]
I vote for "At Large", which we could upgrade later.
1. We currently don't have any projects pending.
2. when projects: we would have to concretely define what kind of support we expect of the 
$FOUNDATION. And decide wether we have the time resources to satisfy their criterias.

**Comment by boneskull** [comment]
@jorydotcom 	Can you please clarify this?

> * Projects that are able to attract a healthy number of committers on the basis of its production usefulness (not simply 'developer popularity').

**Comment by boneskull** [comment]
Upon further reflection: it's unreasonable to adopt "LTS" *without* a release cadence.  And I do believe an LTS offering would be in the project's best interests.  The timelines of such a thing are TBD, and I'm sure we could arrive at something that works for us.

so...

I feel like I'd like Mocha to remain in the "growth" stage.  Without more discussion around specific project goals and steps to reach them, it seems silly to discount the idea wholesale.  They may very well be achievable!  If we determine that the expectations are unreasonable, we can move out of "growth" and into "at large".  That might be sooner rather than later, if the initial discussions don't pan out.

In summary, I think we should test the waters of "growth" under the new foundation.  While we *could* move from "at large" to "growth", it will surely prove more difficult than moving from "growth" to "at large" if need be.

@juergba @craigtaub 	thoughts?

**Comment by jorydotcom** [comment]
> @jorydotcom Can you please clarify this?
> 
> > * Projects that are able to attract a healthy number of committers on the basis of its production usefulness (not simply 'developer popularity').

I think the intention here is to reflect that there are many many projects in production, that aren't "trendy" (and forgive me if this comes across as a pejorative, it is not intended). Your project doesn't have to be popular on twitter, reddit, hackernews, etc in order to have an impact.


Also, regarding moving from one stage to another - it should be easy to move from one stage to another if you meet the requirements of that stage - there is no rank/hierarchy meant to be implied with these stages.

**Comment by juergba** [comment]
My thoughts:
- we are already in "Growth" (~"incubator"). Will the presence of the new FOUNDATION be different, more perceivable than before, in case we just remain in "Growth"?
- "LTS" makes sense yes, but this requirement is already "Impact". With our Pre-Release, we have made a first step towards "LTS".
> Growth - for projects that have identified growth goals, and want active foundation support to reach them.

- Our decision taking should start at this end. Some days ago you removed the "projects" section out of Mocha's repo. This discussion reminds me of ISO certification processes, where you are required to do strange things in order to get a title.
- for the user: does it make a difference wether we are "At Large" or "Growth"?
- we could just cherry pick out of the wall of text, what seems to be useful and we have time for.
 - I can easily live with "Growth"

**Comment by craigtaub** [comment]
Your right we can always move _back_ to "at large" (it could happen to us after a year anyway).

Lets plan to discuss some project/growth goals, if we come up with nothing worthwhile for us then we will have our answer.

Im assuming with LTS we could apply from current release onwards, in which case no major op changes at this point (im completely inexperienced here tho so we could probably do with a discussion on that too).

"Growth" sounds good to me 👍

**Comment by boneskull** [comment]
@juergba 

> * we are already in "Growth" (~"incubator"). Will the presence of the new FOUNDATION be different, more perceivable than before, in case we just remain in "Growth"?

Hopefully this will be more "perceivable", yes, as the merged foundation intends to have more resources to dedicate to growth-phase projects.  I don't know what that will actually look like... but that's the idea.

> * "LTS" makes sense yes, but this requirement is already "Impact". With our Pre-Release, we have made a first step towards "LTS".

This isn't what "LTS" implies (though this is not defined anywhere, unfortunately). In a nutshell, what it'd mean for us is that we commit to supporting a major version of Mocha for some defined period of time, regardless of whether or not we release another major version.  For _example_, we could commit to backporting fixes to the v5.x line of Mocha for another *n* months from now.

> * Our decision taking should start at this end. Some days ago you removed the "projects" section out of Mocha's repo. This discussion reminds me of ISO certification processes, where you are required to do strange things in order to get a title.

I removed GitHub "projects" from mochajs/mocha because it was unused, and the information in it (if any) was outdated.  We can re-open it again if we want to use it.

> * for the user: does it make a difference wether we are "At Large" or "Growth"?

In "growth", the user would naturally see any benefits that arise because of this project working towards its goals.  Otherwise, no...

> * we could just cherry pick out of the wall of text, what seems to be useful and we have time for.

I'm not sure what this is referring to.

> Lets plan to discuss some project/growth goals, if we come up with nothing worthwhile for us then we will have our answer.

I think what @jorydotcom needs from us (please correct me if I'm wrong) is an *intent* to put ourselves in one of these buckets.  Once we give the green light to Mocha being a "growth phase" project, then we can take next steps towards formalizing goals, including what our LTS would look like.

**Comment by cspotcode** [comment]
I'm not a maintainer, but I have a couple ideas, after looking at this thread and the discussion in https://gitter.im/mochajs/contributors.  This is less about JSF and more about mocha's development process.

1. Oddly enough, all the money-related imagery on mocha's README has been a turn-off from contributing for the past few years.  (talk of "backers" and "donations")  I should preface this by saying it may not be a rational reaction.  But when I get the strong impression that other people are being paid to develop mocha, I'm far less inclined to send pull requests.  I'd rather make an issue and hope the paid developers implement it.  I think -- whether it makes sense or not -- this feels more "fair" to me.  I feel like an outsider, so I'd rather leave the development to the insiders who are perhaps more connected to the project's design goals.  In saying this, I realize that donations are a good thing, and paying people for their time is a good thing.  I'm not sure how to reconcile the two.

2. For empowering other maintainers to cut new releases: would it help to formalize the release process in CI/CD?  The [checklist](https://github.com/mochajs/mocha/blob/master/MAINTAINERS.md#mochas-release-process) is intimidating.  CI/CD could theoretically run automatic verifications to ensure master is ready to publish a new release. (changelog updated, tagged correctly, etc) and then have a manually-triggered step to do the actual publication.

Thanks!

**Comment by jorydotcom** [comment]
> I think what @jorydotcom needs from us (please correct me if I'm wrong) is an _intent_ to put ourselves in one of these buckets. Once we give the green light to Mocha being a "growth phase" project, then we can take next steps towards formalizing goals, including what our LTS would look like.

that's right - you just need to let us know what stage you intend to join for now, and as we bootstap the official CPC in the new foundation you may discover that another phase is a better fit, which is a-ok... you'll be able to move if you want/need. We'll be really shaping and defining more clearly what the stages mean and the value they provide as we go (and we'll be doing that together)

**Comment by boneskull** [comment]
@jorydotcom sounds like those of us who are interested in participating feel “growth” stage is workable for now then.

**Comment by boneskull** [comment]
@cspotcode Mocha has donations around something like $20k USD/year, and no maintainers are paid from that anyhow.

**Comment by jorydotcom** [comment]
Sounds good - thanks for the discussion, everyone 😄
